Maya: Add optional Fbx extractors in Rig and Animation family

moonyuet opened the pull request:
## Changelog Description
This PR allows user to export control rigs(optionally with mesh) and animated rig in fbx optionally by attaching the rig objects to the two newly introduced sets.

### Rig Family
Two new sets are introduced in the Rig family: `skeletonMesh_SET` and `skeletonAnim_SET` 
`skeletonMesh_SET`: User can optionally attach the control rigs and/or mesh without animation to the set `skeletonMesh_SET` 
and the objects in the set would be published and exported as fbx
`skeletonAnim_SET`: User can optionally attach the control rigs and it will be published along with the other sets in the original rig extractor
### Animation Family
`skeletonAnim_SET`: User can optionally publish and export the set as fbx with the baked animated data. 

### Validators
`ValidateSkeletonRigContents`: Similar to` Validate Rig Contents`. 
Make sure the rigs optionally contain at least two object sets: `skeletonAnim_SET `- Set of only bone hierarchies and `skeletonMesh_SET` - Set of all cacheable meshes

`ValidateSkeletonRigControllers`: Similar to `ValidateRigControllers`. Validate rig controller for `skeletonAnim_SET ` to make sure all controllers the transformation attributes on their default values of translate zero, rotate zero and scale one when they are unlocked attributes.

`ValidateSkeletonRigOutSetNodeIds`: Similar to `ValidateRigOutSetNodeIds`. Validate if deformed shapes have related IDs to the original shapes from skeleton set.

`ValidateSkeletonRigOutputIds`: Similar to `ValidateRigOutputIds `. Validate rig output ids from the skeleton sets.

## Additional info
There are rooms for discussions on the names of the sets.
It is still mandatory to include the rig objects in both `out_set` and `controls_set`
There is a validator bug fix in #5595, it is affecting all the validators here.

## Testing notes:
Rig Family:
1. Launch Maya via launcher
2. Create Rig Instance
3. Attach your rigs and geometries to out_set & controls_set as usual
4. Attach your control rigs and/or meshes into `skeletonMesh_SET`
5. Attach your control rigs into  `skeletonAnim_SET` 
6. Publish

Animation Family:
1. Doing all the steps with "Rig Family" testing notes above
2. Load your rigs 
3. Animate your rigs
4. Click "Publish...." via Menu
5. Publish along with the animation family

BigRoy commented on openpype/hosts/maya/plugins/publish/extract_rig_fbx.py:
```diff
@@ -0,0 +1,92 @@
+            self.log.debug(
+                "Extract animated FBX successful to: {0}".format(path))
+
+    def _to_extract(self, instance, path, members):
```
Why is this called `_to_extract`? :) Why not just `extract`?

BigRoy commented on openpype/hosts/maya/plugins/publish/extract_rig_fbx.py:
```diff
@@ -0,0 +1,92 @@
+
+        # Export
+        with maintained_selection():
+            fbx_exporter.export(members, path)
```
If the `FBXExtractor` class allows to specify the members to export why do we still need to select the members?

BigRoy commented on openpype/hosts/maya/plugins/publish/extract_rig_fbx.py:
```diff
@@ -0,0 +1,92 @@
+            instance.data["animationOnly"] = True
+
+        fbx_exporter.set_options_from_instance(instance)
+
+        # Export
+        with maintained_selection():
+            fbx_exporter.export(members, path)
+            cmds.select(members, r=1, noExpand=True)
+            mel.eval('FBXExport -f "{}" -s'.format(path))
```
The call to `fbx_exporter.export` above should already be doing an export, right?

BigRoy replied:
```suggestion

        # Export
        fbx_exporter.export(members, path)
```
Move maintained selection into the `FBXExtractor.export` method.

BigRoy commented on openpype/hosts/maya/plugins/publish/extract_rig_fbx.py:
```diff
@@ -0,0 +1,92 @@
+        fbx_exporter.set_options_from_instance(instance)
+
+        # Export
+        with maintained_selection():
```
Let's build the `maintained_selection` behavior into the `FBXExtractor.export` method since you'd not expect that to change selection either.

BigRoy commented on openpype/hosts/maya/plugins/publish/extract_rig_fbx.py:
```diff
@@ -0,0 +1,92 @@
+        if members == skeletal_mesh:
+            instance.data["constraints"] = True
+            instance.data["referencedAssetsContent"] = True
+            instance.data["animationOnly"] = True
```
I feel it's a bit weird that WHEN a skeletal mesh is present only animation is ever exported. Since this applies to the `rig` family and not to the `animation` family (which is actually what is animated with a rig) it doesn't make much sense?

BigRoy commented on openpype/hosts/maya/plugins/publish/extract_rig_fbx.py:
```diff
@@ -0,0 +1,92 @@
+        if skeletal_mesh:
+            self._to_extract(instance, path, skeletal_mesh)
+            representation = {
+                'name': 'fbxanim',
```
Why can't this just be `fbx`? Is there another fbx representation we're expecting for a rig. And when does a `rig` publish actually contain animation - that seems wrong. `animation` family publishes what is the resulting animation made with a loaded rig.

BigRoy reviewed (commented):
Some things that stood out to me.

LiborBatek commented:
@BigRoy  just to let u keep on track, what we are striving to get here being:

```skeleton``` eventually ```skeleton_anim``` only containing baked bones hiearchy animation in FBX (as animation asset for realtime usage like UE, Unity etc)
```skeleton_mesh``` containing skeleton with skinned meshes, no animation, just rig pose (intended as an skeletalMesh asset for use with skeleton anim assets for animations)

both originating from ```rig``` family aka instance in the scene containing following sets:

**Rig**
   **- geo_set**          (currently "out_set")
   **- rig_set**            (currently "control_set")
   **- skeleton_set**   (new one, producing those aformetioned products)

The naming not decided yet so keep it with reserve.

BigRoy commented:
> eventually skeleton_anim only containing baked bones hiearchy animation in FBX (as animation asset for realtime usage like UE, Unity etc)

But those would be coming from the loaded rig from an animation family, right? Since the rig isn't animated? :) Unless you mean you're exporting a single frame 'example cache' which seems redundant since the bones are included in the other export too?

LiborBatek commented:
@BigRoy  ```skeleton_set``` should have bones hiearchy driven by the ```rig_set``` but being completely separate hiearchy (imagine kind of contraints driven bones producing / mimicking the animation from the normal rig...kind of duplicate suitable for realtime engines which supports bones only)

Does it make any sense now? Not super sure you were asking this.

BigRoy commented on openpype/hosts/maya/plugins/publish/validate_rig_contents.py:
```diff
@@ -148,3 +151,63 @@ def validate_controls(self, set_members):
         return invalid
+
+    def validate_skeleton_sets(self, instance):
+        objectsets = ["skeletonAnim_SET", "skeletonMesh_SET"]
+        missing = [obj for obj in objectsets if obj not in instance]
+        if missing:
+            self.log.debug("%s is missing %s" % (instance, missing))
```
Are these optional or required? I believe they are optional. Right?
Also, should we put this validation for skeleton rig sets into a validator of its own? It's a relatively specific validator. It's best to have separate dedicated validators so we can also create the best consistent client facing reports for each validation.

```suggestion
        required = ["skeletonAnim_SET", "skeletonMesh_SET"]
        missing = [
            key for key in required if key not in instance.data["rig_sets"]
        ]
        if missing:
            self.log.debug("%s is missing %s" % (instance, missing))
            return
```

BigRoy commented on openpype/hosts/maya/plugins/publish/validate_rig_controllers.py:
```diff
@@ -76,7 +76,20 @@ def get_invalid(cls, instance):
                 "All controls must be inside the rig's group."
             )
             return [controls_set]
+        skeleton_set = instance.data["rig_sets"].get("skeletonAnim_SET")
+        if not skeleton_set:
+            cls.log.info(
+                "No 'skeletonAnim_SET' in rig instance"
+            )
+        skeleton_controls = cmds.sets(skeleton_set, query=True)
+        if not all(control in lookup for control in cmds.ls(skeleton_controls,
+                                                            long=True)):
+            cls.log.error(
+                "All controls must be inside the rig's group."
+            )
+            return [skeleton_controls]
```
Should we move this into a separate validator `validate_rig_skeleton_anim` since it's basically isolated from the regular controls for a rig, right?

Also, if the `skeleton_set` does not exist then you couldn't continue to query the `cmds.sets` for it so this would error.

BigRoy commented on openpype/hosts/maya/plugins/publish/validate_rig_out_set_node_ids.py:
```diff
@@ -58,7 +62,8 @@ def get_invalid(cls, instance):
                          noIntermediate=True)
-
+        if not shapes:
+            return
```
```suggestion
            return []
```

BigRoy commented on openpype/hosts/maya/plugins/publish/validate_rig_output_ids.py:
```diff
@@ -53,7 +57,20 @@ def get_invalid_matches(cls, instance, compute=False):
+            if not skeletonMesh_set:
+                instance.data["mismatched_output_ids"] = invalid
+                return invalid
+            else:
+                skeletonMesh_nodes = cmds.sets(
+                    skeletonMesh_set, query=True, nodesOnly=True)
+                instance_nodes += skeletonMesh_nodes
```
This doesn't work - the `compute` toggle won't work for the different sets because there's only one `mismatched_output_ids` cache, etc. 

This should be done differently and you're likely best of initially already combining all members you need to query together.

If there's no skeleton mesh set this will now never validate and always return `{}`

moonyuet replied:
I will try to put all these changes into separate validators

BigRoy commented on openpype/hosts/maya/plugins/publish/validate_rig_output_ids.py:
```diff
@@ -53,7 +57,20 @@ def get_invalid_matches(cls, instance, compute=False):
+
             instance_nodes = cmds.ls(instance_nodes, long=True)
+            if not instance_nodes:
+                return
```
```suggestion
                return {}
```

BigRoy commented on openpype/hosts/maya/plugins/publish/validate_skeleton_rig_content.py:
```diff
@@ -0,0 +1,138 @@
+    """
+
+    order = ValidateContentsOrder
+    label = "Rig Contents"
```
Make sure to update the labels so it's clear that it's solely about the "skeleton rig" part.
Also, could we target this validators using `families` so they are only ever run when the instance is actually expected to need/have skeleton meshes?

Like in the Collect Rig For FBX plugin we e.g. append a family like `rig.fbx` or alike so that we target all of these validators to `rig.fbx` family?

BigRoy commented on openpype/hosts/maya/plugins/publish/validate_skeleton_rig_content.py:
```diff
@@ -0,0 +1,138 @@
+
+    def process(self, instance):
+
+        objectsets = ["skeletonAnim_SET", "skeletonMesh_SET"]
+        missing = [obj for obj in objectsets if obj not in instance]
+        if missing:
+            self.log.debug("%s is missing %s" % (instance, missing))
+            return
```
```suggestion
        objectsets = ["skeletonAnim_SET", "skeletonMesh_SET"]
        missing = [
            key for key in objectsets if key not in instance.data["rig_sets"]
        ]
        if missing:
            self.log.debug(
                "%s is missing sets: %s" % (instance, ", ".join(missing))
            )
            return
```

BigRoy commented on openpype/plugins/publish/integrate.py:
```diff
@@ -105,6 +105,7 @@ class IntegrateAsset(pyblish.api.InstancePlugin):
                 "review",
                 "rendersetup",
                 "rig",
+                "rig.fbx",
```
```suggestion
```
This is not needed - you are only appending that family for the publish process but the `rig` family is still present, and thus it'd still be integrated.

LiborBatek commented:
Gives me error during ```Publish``` on ```ExtractFBX``` even tho ```skeletonAnim``` and ```skeletonMesh``` sets being empty.

```
INFO: Export options: {'cameras': False, 'smoothingGroups': True, 'hardEdges': False, 'tangents': False, 'smoothMesh': True, 'instances': False, 'bakeComplexAnimation': True, 'bakeComplexStart': 1001, 'bakeComplexEnd': 1050, 'bakeComplexStep': 1, 'bakeResampleAnimation': True, 'animationOnly': True, 'useSceneName': False, 'quaternion': 'euler', 'shapes': True, 'skins': True, 'constraints': True, 'lights': True, 'embeddedTextures': False, 'inputConnections': True, 'upAxis': 'y', 'triangulate': False, 'exportFileVersion': 'FBX201000'}
INFO: FBXExportCameras -v false
INFO: FBXExportSmoothingGroups -v true
INFO: FBXExportHardEdges -v false
INFO: FBXExportTangents -v false
INFO: FBXExportSmoothMesh -v true
INFO: FBXExportInstances -v false
INFO: FBXExportBakeComplexAnimation -v true
INFO: FBXExportBakeComplexStart -v 1001
INFO: FBXExportBakeComplexEnd -v 1050
INFO: FBXExportBakeComplexStep -v 1
INFO: FBXExportBakeResampleAnimation -v true
INFO: FBXExportAnimationOnly -v true
INFO: FBXExportUseSceneName -v false
INFO: FBXExportQuaternion -v euler
INFO: FBXExportShapes -v true
INFO: FBXExportSkins -v true
INFO: FBXExportConstraints -v true
INFO: FBXExportLights -v true
INFO: FBXExportEmbeddedTextures -v false
INFO: FBXExportInputConnections -v true
INFO: FBXExportUpAxis y
INFO: FBXExportTriangulate -v false
INFO: FBXExportExportFileVersion -v FBX201000
Traceback (most recent call last):
  File "D:\REPO\OpenPype\.venv\lib\site-packages\pyblish\plugin.py", line 527, in __explicit_process
    runner(*args)
  File "D:\REPO\OpenPype\openpype\hosts\maya\plugins\publish\extract_fbx_animation.py", line 43, in process
    fbx_exporter.set_options_from_instance(instance)
  File "D:\REPO\OpenPype\openpype\hosts\maya\api\fbx.py", line 188, in set_options_from_instance
    mel.eval(cmd)
RuntimeError: Error occurred during execution of MEL script
line 1: line 1: Cannot find procedure "FBXExportExportFileVersion".
```
![image](https://github.com/ynput/OpenPype/assets/112623825/b9f3a6b9-40d1-4b53-939f-3cd3fc4b488f)

moonyuet commented:
> Gives me error during `Publish` on `ExtractAnimation (FBX)` even tho `skeletonAnim` and `skeletonMesh` sets being empty.
> 
> ```
> INFO: Export options: {'cameras': False, 'smoothingGroups': True, 'hardEdges': False, 'tangents': False, 'smoothMesh': True, 'instances': False, 'bakeComplexAnimation': True, 'bakeComplexStart': 1001, 'bakeComplexEnd': 1050, 'bakeComplexStep': 1, 'bakeResampleAnimation': True, 'animationOnly': True, 'useSceneName': False, 'quaternion': 'euler', 'shapes': True, 'skins': True, 'constraints': True, 'lights': True, 'embeddedTextures': False, 'inputConnections': True, 'upAxis': 'y', 'triangulate': False, 'exportFileVersion': 'FBX201000'}
> INFO: FBXExportCameras -v false
> INFO: FBXExportSmoothingGroups -v true
> INFO: FBXExportHardEdges -v false
> INFO: FBXExportTangents -v false
> INFO: FBXExportSmoothMesh -v true
> INFO: FBXExportInstances -v false
> INFO: FBXExportBakeComplexAnimation -v true
> INFO: FBXExportBakeComplexStart -v 1001
> INFO: FBXExportBakeComplexEnd -v 1050
> INFO: FBXExportBakeComplexStep -v 1
> INFO: FBXExportBakeResampleAnimation -v true
> INFO: FBXExportAnimationOnly -v true
> INFO: FBXExportUseSceneName -v false
> INFO: FBXExportQuaternion -v euler
> INFO: FBXExportShapes -v true
> INFO: FBXExportSkins -v true
> INFO: FBXExportConstraints -v true
> INFO: FBXExportLights -v true
> INFO: FBXExportEmbeddedTextures -v false
> INFO: FBXExportInputConnections -v true
> INFO: FBXExportUpAxis y
> INFO: FBXExportTriangulate -v false
> INFO: FBXExportExportFileVersion -v FBX201000
> Traceback (most recent call last):
>   File "D:\REPO\OpenPype\.venv\lib\site-packages\pyblish\plugin.py", line 527, in __explicit_process
>     runner(*args)
>   File "D:\REPO\OpenPype\openpype\hosts\maya\plugins\publish\extract_fbx_animation.py", line 43, in process
>     fbx_exporter.set_options_from_instance(instance)
>   File "D:\REPO\OpenPype\openpype\hosts\maya\api\fbx.py", line 188, in set_options_from_instance
>     mel.eval(cmd)
> RuntimeError: Error occurred during execution of MEL script
> line 1: line 1: Cannot find procedure "FBXExportExportFileVersion".
> ```
> 
> ![image](https://user-images.githubusercontent.com/112623825/267035295-b9f3a6b9-40d1-4b53-939f-3cd3fc4b488f.png)

it seems that this parm of determining the fbx version is no longer being used. I will remove the parameter then

LiborBatek commented:
When adding bones hiearchy into ```skeletonAnim_SET``` it gives me following error:
![image](https://github.com/ynput/OpenPype/assets/112623825/6b65e3db-73d8-4236-8017-b5b5c641892c)

moonyuet commented:
> When adding bones hiearchy into `skeletonAnim_SET` it gives me following error: ![image](https://user-images.githubusercontent.com/112623825/267037723-6b65e3db-73d8-4236-8017-b5b5c641892c.png)

I have set the wrong families.(Dumb mistake....)

BigRoy commented on openpype/hosts/maya/plugins/publish/extract_rig_fbx.py:
```diff
@@ -0,0 +1,64 @@
+    """
+    order = pyblish.api.ExtractorOrder
+    label = "Extract Rig (FBX)"
+    families = ["rig"]
+
+    def process(self, instance):
+        if not self.is_active(instance.data):
+            return
+        if "rig.fbx" not in instance.data["families"]:
+            self.log.debug("No object inside skeletonMesh_set..Skipping..")
+            return
```
```suggestion
    families = ["rig.fbx"]

    def process(self, instance):
        if not self.is_active(instance.data):
            return
```

moonyuet commented:
@LiborBatek can you do a test with this updated version?

LiborBatek commented:
When Publish the ```rig``` there arent any FBX files produced when ```skeletonMesh_set``` filled with objects (skeletonAnim left empty on purpose)

When trying to put bones hiearchy (aka skeleton) inside ```skeletonAnim_set``` I got following error:

![image](https://github.com/ynput/OpenPype/assets/112623825/f301c8cb-30bd-408a-ae7e-b9011f0f9d3c)

antirotor commented on openpype/hosts/maya/plugins/publish/collect_fbx_animation.py:
```diff
@@ -0,0 +1,30 @@
+    label = "Collect Fbx Animation"
+    families = ["animation"]
+
+    def process(self, instance):
+        frame = cmds.currentTime(query=True)
+        instance.data["frameStart"] = frame
+        instance.data["frameEnd"] = frame
```
I am not entirely sure about this - shouldn't this collect frame range and not the single frame? If there is some specific reason, I would put it in the comment.

antirotor reviewed (commented):
Just a question about frame range

LiborBatek commented:
This time not any errors during ```Publish``` of the rig but no any ```FBX``` product written...
![image](https://github.com/ynput/OpenPype/assets/112623825/94370c3b-6cf1-4fea-9de4-8fa9de5b9b9e)

moonyuet commented:
> This time not any errors during `Publish` of the rig but no any `FBX` product written... ![image](https://user-images.githubusercontent.com/112623825/267903587-94370c3b-6cf1-4fea-9de4-8fa9de5b9b9e.png)

It could be because it doesn't include maya as host in the extractor. I just made the commit to update maya as hosts.

antirotor commented on openpype/hosts/maya/plugins/publish/collect_fbx_animation.py:
```diff
@@ -0,0 +1,27 @@
+    def process(self, instance):
+        skeleton_sets = [
+            i for i in instance
+            if i.lower().endswith("skeletonanim_set")
```
I don't think this should ever happen. What is the source of `skeletonamin_set`?

antirotor replied:
because that is the current cause of why **Extract Animation (FBX)** never triggers

moonyuet replied:
> I don't think this ever happen. What is the source of `skeletonamin_set`?

The `skeletonamin_set` should be skeletonAnim_SET. So this code is to get the list of selection from skeletonAnim_SET

moonyuet replied:
I have figured out why, the anim_skeleton set and skeleton_mesh set have not been added into the loaded rig sets so that's why the animated_skeleton got empty list

LiborBatek commented:
So now it finally produces both repre for the ```rig``` native .ma and ```Fbx``` one.

![image](https://github.com/ynput/OpenPype/assets/112623825/c12f3304-8f26-410d-b80b-c5bbc6d344eb)

Question being how to treat those when using ```Loader``` because it fails on missing rig sets (out_Set and control_set) when using ```Reference as FBX```
![image](https://github.com/ynput/OpenPype/assets/112623825/14cc81fe-ff03-4428-9283-fa967a590bb7)

![image](https://github.com/ynput/OpenPype/assets/112623825/9bffb692-f5ac-41ff-a049-c5eb9f379f49)

however its loaded into scene...and working.

LiborBatek reviewed (changes requested):
Publishing of the ```rig``` correctly produces also ```Fbx``` representation of the ```sekeletonMesh``` set.

There are however issues with ```skeletonAnim``` set which incorrectly publishes animation in ```Fbx``` containing invalid date...see scrngrb when loaded into scene as asset:

![image](https://github.com/ynput/OpenPype/assets/112623825/839f1a1e-954e-43ab-b632-e36bd102e052)

Pls. Note that the visible mesh represents just the normal ```animation``` product in ```alembic``` and that we completely miss ```skeletonAnim``` because of invalid datas inside...

moonyuet commented:
> Publishing of the `rig` correctly produces also `Fbx` representation of the `sekeletonMesh` set.
> 
> There are however issues with `skeletonAnim` set which incorrectly publishes animation in `Fbx` containing invalid date...see scrngrb when loaded into scene as asset:
> 
> ![image](https://user-images.githubusercontent.com/112623825/268607051-839f1a1e-954e-43ab-b632-e36bd102e052.png)
> 
> Pls. Note that the visible mesh represents just the normal `animation` product in `alembic` and that we completely miss `skeletonAnim` because of invalid datas inside...
> 
> Also the produced `animation` data in `Fbx` should have probably slightly different naming I guess. Being it now as following: ![image](https://user-images.githubusercontent.com/112623825/268610820-ebc05373-57d7-40d8-90ec-4b3fca2adbb1.png)

@LiborBatek I have fixed the animation export issue. (It shouldn't be exporting the transform attribute anymore). I also added the fbx loader for animation family and fixed mandatory loading of out_set with the fbx loader in rig family.

LiborBatek commented:
So now OP correctly produces ```animation``` in both native ```maya``` repre and ```Fbx``` with just ```skeletonAnim``` containing just skeleton hiearchy.
![image](https://github.com/ynput/OpenPype/assets/112623825/9684e498-6cd4-4820-baeb-760434b92f2a)

question being if we should keep ```Namespaces``` during ```Publish``` or strip em off (speaking of UE) Not sure about this one, namely regarding OP practices we use for native ```animation``` family.

![image](https://github.com/ynput/OpenPype/assets/112623825/eade967a-884c-4e80-8a18-1a22034629aa)

another question being if there shouldnt be any switch (optional knob) regarding outputing ```skeletonAnim``` when ```Publish Animation``` ...right now it happens automatically each time.

![image](https://github.com/ynput/OpenPype/assets/112623825/206f1fa4-3262-48de-89a9-a60d51997c09)

LiborBatek commented:
there is also inconsistency when ```Publishing```  ```Animation``` now.... there is not ```animation``` prefix present on the ```animation``` instance. Compare this with what we have now:
![image](https://github.com/ynput/OpenPype/assets/112623825/0a0b3fab-8c5f-4fd3-afdb-902ee14938a1)

LiborBatek commented:
So regarding the stripping ```Namespaces``` its clear now...we need to produce ```skeletonAnim``` with stripped namespaces so the

```skeletonMesh``` asset matches our ```animation```

correctly publish ```skeletonMesh``` has stripped namespaces (preferable) and animation dont (here just manually imported to maya to check how the skeletonMesh looks inside naming wise) ....this being correct btw

![image](https://github.com/ynput/OpenPype/assets/112623825/e3c8d03b-3782-4479-9471-cde92db0f095)

We need to strip namespaces on ```animSkeleton``` too when producing ```FBX```

LiborBatek commented:
and one last thing...we need to ensure that both assets

```skeletonMesh``` - correct hiearchy
```skeletonAnim``` - one more node above

having the same hiearchy which they dont right now (as being inspected by manually importing FBX into maya from published data. This I think being UE mandatory to be able to connect those aka apply ```skeletonAnim``` on ```skeletonMesh```

see the difference here (compare with pic from preceding posts)

![image](https://github.com/ynput/OpenPype/assets/112623825/3fa28fb0-6f0e-4060-a9e4-36c243022f8e)

moonyuet commented:
@LiborBatek can you please test again?

LiborBatek commented:
Did another testing go and now ```Namespaces``` being correctly stripped...the only problem being there is top ```_GRP``` node above the desired ```rig_GRP``` still.

This being caused by origin of the ```animation``` where ```rig``` asset being loaded and the top ```GRP``` automatically created as seen on the following pic

![image](https://github.com/ynput/OpenPype/assets/112623825/0b6b4a0e-06df-4360-812e-f7f052275e21) and after stripping the Namespaces:
![image](https://github.com/ynput/OpenPype/assets/112623825/38b08ec0-180c-4d8a-a96d-bfa60325f871)

And lastly the ```skeletalMesh``` itself as published from ```maya``` 

![image](https://github.com/ynput/OpenPype/assets/112623825/848cca2f-f3e8-42de-8bcd-e8968ceaf3d7)

Question being how to treat the top group which got appended inside ```animation``` workfile after asset load. There is still some possibility user wont use grouping when loading the asset which would result into correct skeletalAnim hiearchy...

![image](https://github.com/ynput/OpenPype/assets/112623825/c3a29139-de5d-4706-bbea-c568ebc89f12)

LiborBatek commented:
Did test to output ```animation``` from local rig without referencing, the validator correctly kicks in but ```Select Invalid``` btn does not work at all....not sure if caused within this PR or being issue in general in OP.
![image](https://github.com/ynput/OpenPype/assets/112623825/3aa1257a-cb72-411f-aa54-22a52d3041a0)

Minkiu commented on openpype/hosts/maya/plugins/publish/collect_fbx_animation.py:
```diff
@@ -0,0 +1,33 @@
+            if i.lower().endswith("skeletonanim_set")
+        ]
+        if skeleton_sets:
+            instance.data["families"] += ["animation.fbx"]
```
Why not `instance.data["families"].append("animation.fbx")` ?

moonyuet replied:
i agree it's the same.

Minkiu commented on openpype/hosts/maya/plugins/publish/extract_fbx_animation.py:
```diff
@@ -0,0 +1,68 @@
+        cmds.namespace(set=':')
+        cmds.namespace(rel=False)
+
+        if "representations" not in instance.data:
```
This could be:
```py
representations = instance.data.setdefault("representations", [])
representations.append({
            'name': 'fbx',
            'ext': 'fbx',
            'files': filename,
            "stagingDir": staging_dir
})
```

moonyuet replied:
i didn't change this t as we need to extract animation in abc  before extract fbx animation.

BigRoy replied:
The recommended change by @Minkiu would behave exactly the same as your current logic.

`dict.setdefault` set it to the value only if it does not exist, else it returns the value - and thus you'd actually be appending to the `instance.data["representations"]` that already existed. So the example code provided is still valid even for your use case.

LiborBatek commented:
Did another testing round with local rig asset (non referenced asset) and the ```Select Invalid``` btn still not selecting anything...its minor thing but not much of a use now...
![image](https://github.com/ynput/OpenPype/assets/112623825/67c5ffb3-3c60-442a-8604-58e2a0f2905d)

LiborBatek reviewed (approved):
All seems working fine, now when no referenced asset present, just error msg shows up without any repair aka select invalid nodes btn.

There are no issues at this time present as far as we are aware of neccessary functionality of this particular PR.

BigRoy commented on openpype/hosts/maya/api/lib.py:
```diff
@@ -4100,15 +4100,24 @@ def create_rig_animation_instance(
+        assert output, "No out_SET in rig, this is a bug."
+        assert controls, "No controls_SET in rig, this is a bug."
+
+    anim_skeleton = next((node for node in nodes if
+                          node.endswith("skeletonAnim_SET")), None)
+    if not anim_skeleton:
+        log.debug("No skeletonAnim_SET in rig")
+
+    skeleton_mesh = next((node for node in nodes if
+                          node.endswith("skeletonMesh_SET")), None)
+    if not skeleton_mesh:
+        log.debug("No skeletonMesh_SET in rig")
```
So the sets above do not exist for an `fbx` - but these do?
Also, do we care enough to be debug logging these?

moonyuet replied:
if we dont add this, i guess the fbx import would error out to load just the skeletonMesh set.

BigRoy commented on openpype/hosts/maya/api/lib.py:
```diff
@@ -4120,8 +4129,9 @@ def create_rig_animation_instance(
         formatting_data = {
             # TODO remove 'asset_type' and replace 'asset_name' with 'asset'
-            "asset_name": context['asset']['name'],
-            "asset_type": context['asset']['type'],
+            # "asset_name": context['asset']['name'],
+            # "asset_type": context['asset']['type'],
```
So does this now mean we should remove this comment and the commented lines?

BigRoy commented on openpype/hosts/maya/plugins/create/create_rig.py:
```diff
@@ -20,6 +20,13 @@ def create(self, subset_name, instance_data, pre_create_data):
         instance_node = instance.get("instance_node")
 
         self.log.info("Creating Rig instance set up ...")
+        # change name (_controls_set -> _rigs_SET)
         controls = cmds.sets(name=subset_name + "_controls_SET", empty=True)
+        # change name (_out_SET -> _geo_SET)
```
These comments seem odd? Are these `# TODO`?
Also, this'd be a major backwards compatibility hazard ;)
Better leave these alone?

BigRoy commented on openpype/hosts/maya/plugins/publish/collect_fbx_animation.py:
```diff
@@ -0,0 +1,33 @@
+            i for i in instance
+            if i.lower().endswith("skeletonanim_set")
+        ]
+        if skeleton_sets:
+            instance.data["families"].append("animation.fbx")
+            instance.data["animated_skeleton"] = []
+            for skeleton_set in skeleton_sets:
+                skeleton_content = cmds.sets(skeleton_set, query=True)
+                self.log.debug(
+                    "Collected animated "
+                    f"skeleton data: {skeleton_content}")
+                if skeleton_content:
+                    instance.data["animated_skeleton"] += skeleton_content
```
```suggestion
        if not skeleton_sets:
            return
            
        instance.data["families"].append("animation.fbx")
        instance.data["animated_skeleton"] = []
        for skeleton_set in skeleton_sets:
            skeleton_content = cmds.sets(skeleton_set, query=True)
            self.log.debug(
                "Collected animated "
                f"skeleton data: {skeleton_content}")
            if skeleton_content:
                instance.data["animated_skeleton"] += skeleton_content
```

BigRoy commented on openpype/hosts/maya/plugins/publish/collect_skeleton_mesh.py:
```diff
@@ -0,0 +1,52 @@
+            i for i in instance
+            if i.lower().endswith("skeletonmesh_set")
+        ]
+        if not skeleton_sets and skeleton_mesh_sets:
+            self.log.debug(
+                "no skeleton_set or skeleton_mesh set was found....")
+            return
```
Are we sure this logic is correct? This returns only when there are NO skeleton sets but there ARE skeleton mesh sets? Pretty sure this is wrong.

BigRoy commented on openpype/hosts/maya/plugins/publish/collect_skeleton_mesh.py:
```diff
@@ -0,0 +1,52 @@
+    families = ["rig"]
+
+    def process(self, instance):
+        frame = cmds.currentTime(query=True)
+        instance.data["frameStart"] = frame
+        instance.data["frameEnd"] = frame
+        skeleton_sets = [
+            i for i in instance
+            if i.lower().endswith("skeletonanim_set")
+        ]
+
+        skeleton_mesh_sets = [
+            i for i in instance
+            if i.lower().endswith("skeletonmesh_set")
+        ]
+        if not skeleton_sets and skeleton_mesh_sets:
+            self.log.debug(
+                "no skeleton_set or skeleton_mesh set was found....")
+            return
+        instance.data["skeleton_mesh"] = []
+        instance.data["skeleton_rig"] = []
```
```suggestion
        skeleton_set = instance.data.get("skeletonAnim_SET")
        skeleton_mesh_set = instance.data.get("skeletonMesh_SET")
        if not skeleton_sets and not skeleton_mesh_sets:
            self.log.debug(
                "No skeletonAnim_SET or skeletonMesh_SET found. "
                "Skipping collecting of skeleton mesh..."
            )
            return
        
        # Store current frame to ensure single frame export
        frame = cmds.currentTime(query=True)
        instance.data["frameStart"] = frame
        instance.data["frameEnd"] = frame
            
        instance.data["skeleton_mesh"] = []
        instance.data["skeleton_rig"] = []
```
Note: This renames `skeleton_mesh_sets` and `skeleton_sets` to singular names `skeleton_mesh_set` and `skeleton_set`

BigRoy commented on openpype/hosts/maya/plugins/publish/collect_skeleton_mesh.py:
```diff
@@ -0,0 +1,52 @@
+        instance.data["skeleton_mesh"] = []
+        instance.data["skeleton_rig"] = []
+
+        if skeleton_mesh_sets:
+            instance.data["families"] += ["rig.fbx"]
+            for skeleton_mesh_set in skeleton_mesh_sets:
+                skeleton_mesh_content = cmds.sets(
+                    skeleton_mesh_set, query=True)
+                if skeleton_mesh_content:
+                    instance.data["skeleton_mesh"] += skeleton_mesh_content
+                    self.log.debug(
+                        "Collected skeleton "
+                        f"mesh Set: {skeleton_mesh_content}")
+
+        if skeleton_sets:
+            for skeleton_set in skeleton_sets:
+                skeleton_content = cmds.sets(skeleton_set, query=True)
+                self.log.debug(
+                    "Collected animated "
+                    f"skeleton data: {skeleton_content}")
+                if skeleton_content:
+                    instance.data["skeleton_rig"] += skeleton_content
```
What happens if we DO have skeleton_sets but no skeleton_mesh_sets. That's not adding the `rig.fbx` family - are we still expecting that to do something?

moonyuet replied:
maybe mandatory for skeleton_mesh_sets but optiona; for skeleton_set

moonyuet replied:
nothing, i will remove the skeleton_sets

moonyuet replied:
i can remove the skeleton_set to ease the confusion...

BigRoy commented on openpype/hosts/maya/plugins/publish/extract_fbx_animation.py:
```diff
@@ -0,0 +1,65 @@
+        filename = "{0}.fbx".format(instance.name)
+        path = os.path.join(staging_dir, filename)
+
+        # The export requires forward slashes because we need
+        # to format it into a string in a mel expression
```
This comment is in the wrong place since the path swapping doesn't occur here - also, I'd argue that the backslash to forward slash replacing should just happen in the `fbx.FBXExtractor` plugin so whoever uses it doesn't have to care - making this plugin's logic simpler too!

BigRoy commented on openpype/hosts/maya/plugins/publish/extract_fbx_animation.py:
```diff
@@ -0,0 +1,65 @@
+        namespace = out_set_name.split(":")[0]
+        new_out_set = out_set_name.replace(
+            f"{namespace}:", "")
+        cmds.namespace(set=':')
+        cmds.namespace(set=namespace)
```
```suggestion
        cmds.namespace(set=':' + namespace)
```

BigRoy commented on openpype/hosts/maya/plugins/publish/extract_fbx_animation.py:
```diff
@@ -0,0 +1,65 @@
+            f"{namespace}:", "")
+        cmds.namespace(set=':')
+        cmds.namespace(set=namespace)
+        cmds.namespace(rel=True)
```
Use the long argument name here since `rel` is a bit ambiguous when reading it. Assume people don't know maya's API that well that they know all these flags ;)
```suggestion
        cmds.namespace(relativeNames=True)
```

BigRoy commented on openpype/hosts/maya/plugins/publish/extract_fbx_animation.py:
```diff
@@ -0,0 +1,65 @@
+
+        fbx_exporter.export(
+            new_out_set, path.replace("\\", "/"))
+        # restore namespace after export
+        cmds.namespace(set=':')
+        cmds.namespace(rel=False)
```
You are actually **not** restoring the namespace - you're only forcing it to the root namespace afterwards. You might want to use a context manager instead.

I believe `lib` has one for maya:
```python
with namespaced(":" + namespace, new=False) as namespace:
    fbx_exporter.export(new_out_set, path)
```

BigRoy commented on openpype/hosts/maya/plugins/publish/extract_fbx_animation.py:
```diff
@@ -0,0 +1,65 @@
+            "stagingDir": staging_dir
+        })
+
+        self.log.debug("Extract animated FBX successful to: {0}".format(path))
```
```suggestion
        self.log.debug("Extracted animated FBX to: {0}".format(path))
```

BigRoy commented on openpype/hosts/maya/plugins/publish/extract_skeleton_mesh.py:
```diff
@@ -0,0 +1,59 @@
+        }
+        instance.data["representations"].append(representation)
+
+        self.log.debug("Extract animated FBX successful to: {0}".format(path))
```
It's weird that this log message is exactly like the ExtractFBXAnimation one - is this one wrong? :)

BigRoy commented on openpype/hosts/maya/plugins/publish/extract_skeleton_mesh.py:
```diff
@@ -0,0 +1,59 @@
+        # Export
+        fbx_exporter.export(out_set, path.replace("\\", "/"))
+
+        if "representations" not in instance.data:
+            instance.data["representations"] = []
+
+        representation = {
+            'name': 'fbx',
+            'ext': 'fbx',
+            'files': filename,
+            "stagingDir": staging_dir
+        }
+        instance.data["representations"].append(representation)
```
Also use the `representations = instance.data.setdefault("representations", [])` logic here.

BigRoy commented on openpype/hosts/maya/plugins/publish/extract_skeleton_mesh.py:
```diff
@@ -0,0 +1,59 @@
+        fbx_exporter.set_options_from_instance(instance)
+
+        # Export
+        fbx_exporter.export(out_set, path.replace("\\", "/"))
```
See note on other plugin for path slash replacements.

BigRoy commented on openpype/hosts/maya/plugins/publish/validate_animated_reference.py:
```diff
@@ -0,0 +1,63 @@
+
+
+class ValidateAnimatedReferenceRig(pyblish.api.InstancePlugin):
+    """
+        Validate all the nodes underneath skeleton_Anim_SET
+        should be reference nodes
+    """
```
```suggestion
    """Validate all nodes in skeletonAnim_SET are referenced"""
```

Also make sure you always name the sets like you create them in create rig for clarity:
```python
        skeleton = cmds.sets(
            name=subset_name + "_skeletonAnim_SET", empty=True)
        skeleton_mesh = cmds.sets(
            name=subset_name + "_skeletonMesh_SET", empty=True)
```

So this should be `skeletonAnim_SET` instead of `skeleton_Anim_SET`. The same goes for all the logging messages below.

BigRoy commented on openpype/hosts/maya/plugins/publish/validate_animated_reference.py:
```diff
@@ -0,0 +1,63 @@
+        animated_sets = instance.data["animated_skeleton"]
+        if not animated_sets:
+            self.log.debug(
+                "No nodes found in skeleton_Anim_SET..Skipping..")
```
```suggestion
                "No nodes found in skeletonAnim_SET. Skipping..."
            )
```
Make sure to always use the `...` when you're using something like `..` or `....`. Best to be consistent.

BigRoy commented on openpype/hosts/maya/plugins/publish/validate_animated_reference.py:
```diff
@@ -0,0 +1,63 @@
+                animated_reference, isNodeReferenced=True)
+            if not bool(is_referenced):
+                raise PublishValidationError(
+                    "All the content in skeleton_Anim_SET"
+                    " should be reference nodes"
```
```suggestion
                    "All the members of skeletonAnim_SET"
                    " should be referenced nodes"
```

BigRoy commented on openpype/hosts/maya/plugins/publish/validate_animated_reference.py:
```diff
@@ -0,0 +1,63 @@
+                )
+        invalid_controls = self.validate_controls(animated_sets)
+        if invalid_controls:
+            raise PublishValidationError(
+                "All the content in skeleton_Anim_SET"
+                " should be the transforms"
```
```suggestion
            raise PublishValidationError(
                "All the nodes in skeletonAnim_SET"
                " should be transforms"
```

BigRoy commented on openpype/hosts/maya/plugins/publish/validate_skeleton_rig_content.py:
```diff
@@ -0,0 +1,142 @@
+        # in the rig instance
+        set_members = instance.data['setMembers']
+        if not cmds.ls(set_members, type="dagNode", long=True):
+            self.log.debug("Skipping empty instance...")
```
```suggestion
            self.log.debug("Skipping instance without dag nodes...")
```

BigRoy commented on openpype/hosts/maya/plugins/publish/validate_skeleton_rig_content.py:
```diff
@@ -0,0 +1,142 @@
+        if not cmds.ls(set_members, type="dagNode", long=True):
+            self.log.debug("Skipping empty instance...")
+            return
+        # Ensure contents in sets and retrieve long path for all objects
+        skeleton_mesh_content = cmds.sets(
+            skeleton_mesh_set, query=True) or []
+        skeleton_mesh_content = cmds.ls(skeleton_mesh_content, long=True)
+
+        skeleton_anim_content = cmds.sets(
+            skeleton_anim_set, query=True) or []
+        skeleton_anim_content = cmds.ls(skeleton_anim_content, long=True)
```
Didn't we JUST collect these? 🗡️

BigRoy commented on openpype/hosts/maya/plugins/publish/validate_skeleton_top_group_hierarchy.py:
```diff
@@ -0,0 +1,49 @@
+
+    def process(self, instance):
+        invalid = []
+        skeleton_data = instance.data.get(("animated_rigs"), [])
+        skeletonMesh_data = instance.data(("skeleton_mesh"), [])
```
Typos? Are we sure this validator is working and is not listed under "crashed plugins" in the publisher?
```suggestion
        skeleton_data = instance.data.get("animated_rigs"), [])
        skeletonMesh_data = instance.data("skeleton_mesh"), [])
```

BigRoy replied:
Also `skeletonMesh_data` variable name should be `skeleton_mesh_data`?

BigRoy commented on openpype/hosts/maya/plugins/publish/validate_skeleton_top_group_hierarchy.py:
```diff
@@ -0,0 +1,49 @@
+        if skeletonMesh_data:
+            invalid = self.get_top_hierarchy(skeletonMesh_data)
+            if invalid:
+                raise PublishValidationError(
+                    "The set includes the object which "
+                    f"is not at the top hierarchy: {invalid}")
```
What set? Please name the set in the logs. And what is **the** object? Did you mean an object?

BigRoy commented on openpype/hosts/maya/plugins/publish/extract_fbx_animation.py:
```diff
@@ -0,0 +1,64 @@
+        namespace = out_set_name.split(":")[0]
+        new_out_set = out_set_name.replace(
+            f"{namespace}:", "")
+        cmds.namespace(set=':' + namespace)
```
```suggestion
```

moonyuet commented on openpype/hosts/maya/plugins/publish/extract_fbx_animation.py:
```diff
@@ -0,0 +1,64 @@
+            fbx_exporter.export(
+                new_out_set, path.replace("\\", "/"))
+
+        cmds.namespace(relativeNames=False)
```
That's true, have tried in the maya, and it works.

BigRoy commented on openpype/hosts/maya/api/fbx.py:
```diff
@@ -198,5 +203,6 @@ def export(members, path):
         """
-        cmds.select(members, r=True, noExpand=True)
-        mel.eval('FBXExport -f "{}" -s'.format(path))
+        with maintained_selection():
+            cmds.select(members, r=True, noExpand=True)
+            mel.eval('FBXExport -f "{}" -s'.format(path))
```
```suggestion
        # The export requires forward slashes because we need
        # to format it into a string in a mel expression
        path = path.replace("\\", "/") 
        with maintained_selection():
            cmds.select(members, r=True, noExpand=True)
            mel.eval('FBXExport -f "{}" -s'.format(path))
```

BigRoy commented on openpype/hosts/maya/plugins/create/create_rig.py:
```diff
@@ -20,6 +20,13 @@ def create(self, subset_name, instance_data, pre_create_data):
         instance_node = instance.get("instance_node")
 
         self.log.info("Creating Rig instance set up ...")
+        # TODO：change name (_controls_set -> _rigs_SET)
```
Honestly I wouldn't find `_rigs_SET` clearer than `_controls_SET` since it should only include rig controllers so I personally wouldn't recommend this rename.
```suggestion
        # TODO：change name (_controls_SET -> _rigs_SET)
```

BigRoy commented on openpype/hosts/maya/plugins/create/create_rig.py:
```diff
@@ -20,6 +20,13 @@ def create(self, subset_name, instance_data, pre_create_data):
         self.log.info("Creating Rig instance set up ...")
+        # TODO：change name (_controls_set -> _rigs_SET)
         controls = cmds.sets(name=subset_name + "_controls_SET", empty=True)
+        # TODO：change name (_out_SET -> _geo_SET)
```
Same about this name change. The `_out_SET` doesn't necessarily ONLY include geo. It could also have locators that are used in a next step for specific lighting usages.

BigRoy commented on openpype/hosts/maya/plugins/load/_load_animation.py:
```diff
@@ -7,7 +7,7 @@ class AbcLoader(openpype.hosts.maya.api.plugin.ReferenceLoader):
                 "camera",
                 "pointcache"]
-    representations = ["abc"]
+    representations = ["abc", "fbx"]
```
This is a bit weird - the Loader class is called `AbcLoader`, it forces a load of `AbcImport.mll` but for FBX should load `fbx4maya.mll` or whatever the plugin name is. I feel like we're better off with a dedicated FBX loader, no?

Without the FBX plugin loaded this loader will likely fail to load the FBX?

BigRoy commented on openpype/hosts/maya/plugins/publish/collect_fbx_animation.py:
```diff
@@ -0,0 +1,35 @@
+        instance.data["animated_skeleton"] = []
+        for skeleton_set in skeleton_sets:
+            skeleton_content = cmds.sets(skeleton_set, query=True)
+            self.log.debug(
+                "Collected animated "
+                f"skeleton data: {skeleton_content}")
```
f-string formatting is not supported in Py2 and won't work in older maya versions. I believe Maya 2020 is still Py2?

BigRoy commented on openpype/hosts/maya/plugins/publish/collect_skeleton_mesh.py:
```diff
@@ -0,0 +1,39 @@
+                    skeleton_mesh_set, query=True)
+                if skeleton_mesh_content:
+                    instance.data["skeleton_mesh"] += skeleton_mesh_content
+                    self.log.debug(
+                        "Collected skeleton "
+                        f"mesh Set: {skeleton_mesh_content}")
```
f-string formatting is not supported in Py2 and won't work in older maya versions. I believe Maya 2020 is still Py2?

BigRoy commented on openpype/hosts/maya/plugins/publish/extract_fbx_animation.py:
```diff
@@ -0,0 +1,65 @@
+        fbx_exporter.set_options_from_instance(instance)
+
+        out_set_name = next(out for out in out_set)
+        # temporarily disable namespace
```
We're not _disabling_ the namespace but are entering the namespace.
Also, it's unclear WHY we're doing that. Are we doing that because then the FBX exported node names do not include the namespace (basically stripping the namespace in the export?)

```python
# Export from the rig's namespace so that the exported
# FBX does not include the namespace but preserves the node
# names as existing in the rig workfile
```
Something like this would be clearer to me:

BigRoy commented on openpype/hosts/maya/plugins/publish/extract_fbx_animation.py:
```diff
@@ -0,0 +1,65 @@
+
+        out_set_name = next(out for out in out_set)
+        # temporarily disable namespace
+        namespace = out_set_name.split(":")[0]
+        new_out_set = out_set_name.replace(
+            f"{namespace}:", "")
+        cmds.namespace(relativeNames=True)
+        with namespaced(":" + namespace, new=False) as namespace:
+            path = path.replace("\\", "/")
+            fbx_exporter.export(new_out_set, path)
+        original_relative_names = cmds.namespace(
+            query=True, relativeNames=True)
+        if original_relative_names:
+            cmds.namespace(relativeNames=original_relative_names)
```
Please avoid the use of `str.replace()` calls like these. Namespaces can be nested. `:namespace:namespace:namespace:` is technically fine in Maya - what we want to do is just strip of the first that we're entering.

Also, `new_out_set` sounds like you're generating a new node which confuses me a bit.

Additionally you're now querying the `relativeNames` after you've set it so of course it's your value ;) - you want to do that before you're setting it. You're also only setting it IF it's true, but that's not what you want - you want to always revert it to the original value even if it's False.

It should likely be something like this:

```suggestion
        namespace, relative_out_set = out_set_name.split(":", 1)
        original_relative_names = cmds.namespace(
            query=True, relativeNames=True)
        cmds.namespace(relativeNames=True)
        with namespaced(":" + namespace, new=False) as namespace:
            path = path.replace("\\", "/")
            fbx_exporter.export(relative_out_set, path)
        
        cmds.namespace(relativeNames=original_relative_names)
```

This might be made even easier if we add the `relative_names` flag to the `namespaced` context manager so that this code can become:
```python
with namespaced(
    ":" + namespace,
    relative_names=True,
    new=False
) as namespace:
    fbx_exporter.export(relative_out_set, path)
```

Likely even better to maintain like that.

The context manager then would just get an extra argument `relative_names` which defaults to `None` (does nothing), when `False` it forces it to False when `True` it forces it to True and voila. Better code.

BigRoy commented on openpype/hosts/maya/plugins/publish/extract_fbx_animation.py:
```diff
@@ -0,0 +1,65 @@
+        })
+
+        self.log.debug(
+            "Extracted Fbx animation successful to: {0}".format(path))
```
```suggestion
            "Extracted Fbx animation to: {0}".format(path))
```
We hardly know how successful it really was ;) it's just as fine without saying it.

BigRoy commented on openpype/hosts/maya/plugins/publish/extract_skeleton_mesh.py:
```diff
@@ -0,0 +1,57 @@
+        filename = "{0}.fbx".format(instance.name)
+        path = os.path.join(staging_dir, filename)
+
+        # The export requires forward slashes because we need
+        # to format it into a string in a mel expression
```
```suggestion
```

BigRoy commented on openpype/hosts/maya/plugins/publish/extract_skeleton_mesh.py:
```diff
@@ -0,0 +1,57 @@
+        fbx_exporter.set_options_from_instance(instance)
+
+        # Export
+        path = path.replace("\\", "/")
```
```suggestion
```

BigRoy commented on openpype/hosts/maya/plugins/publish/validate_skeleton_top_group_hierarchy.py:
```diff
@@ -0,0 +1,42 @@
+
+    def process(self, instance):
+        invalid = []
+        skeleton_mesh_data = instance.data(("skeleton_mesh"), [])
```
```suggestion
        skeleton_mesh_data = instance.data("skeleton_mesh", [])
```

BigRoy commented on openpype/hosts/maya/plugins/publish/validate_skeleton_top_group_hierarchy.py:
```diff
@@ -0,0 +1,42 @@
+        for target in targets:
+            long_names = cmds.ls(target, long=True)
+            for name in long_names:
+                if len(name.split["|"]) > 2:
```
Pretty sure this is typo - are you sure you sure you've run this code? :)
```suggestion
                if len(name.split("|")) > 2:
```

BigRoy replied:
Also, you don't actually need to split here - you're better off using `str.count`
```suggestion
                if name.count("|") > 2:
```

BigRoy commented on openpype/hosts/maya/plugins/publish/validate_skeleton_top_group_hierarchy.py:
```diff
@@ -0,0 +1,42 @@
+
+    def get_top_hierarchy(self, targets):
+        non_top_hierarchy_list = []
+        for target in targets:
+            long_names = cmds.ls(target, long=True)
```
Doing `ls` queries per node is slow - you can do it for all of them at once
```suggestion
        targets = cmds.ls(targets, long=True)  # ensure long names
        for target in targets:
```

Which also means you don't need the extra for loop but can make it:
```python
        targets = cmds.ls(targets, long=True)  # ensure long names
        for target in targets:
            if target.count("|") > 2:
                non_top_hierarchy_list.append(target)
```
Or if you like list comprehensions:
```python
        targets = cmds.ls(targets, long=True)  # ensure long names
        non_top_hierarchy_list = [
            target for target in targets if target.count("|") > 2
        ]
```

BigRoy commented on openpype/hosts/maya/plugins/publish/validate_skeleton_rig_output_ids.py:
```diff
@@ -0,0 +1,126 @@
+                if shapes:
+                    instance_nodes.extend(shapes)
+
+            scene_nodes = cmds.ls(type="transform", long=True)
+            scene_nodes += cmds.ls(type="mesh", long=True)
```
```suggestion
            scene_nodes = cmds.ls(type=("transform", "mesh"), long=True)
```

BigRoy commented on openpype/hosts/maya/plugins/publish/validate_skeleton_rig_output_ids.py:
```diff
@@ -0,0 +1,126 @@
+        for instance_node, matches in invalid_matches.items():
+            ids = set(get_id(node) for node in matches)
+
+            # If there are multiple scene ids matched, and error needs to be
```
```suggestion
            # If there are multiple scene ids matched, an error needs to be
```

BigRoy commented on openpype/hosts/maya/plugins/publish/validate_skeleton_rig_output_ids.py:
```diff
@@ -0,0 +1,126 @@
+            id_to_set = next(iter(ids))
+            set_id(instance_node, id_to_set, overwrite=True)
+
+        if multiple_ids_match:
+            raise PublishValidationError(
+                "Multiple matched ids found. Please repair manually: "
+                "{}".format(multiple_ids_match)
+            )
```
Are we sure we can raise PublishValidationError reports from a `repair` action. I'm pretty sure the publisher does not support this. @iLLiCiTiT do you know?

BigRoy commented on openpype/hosts/maya/api/lib.py:
```diff
@@ -934,6 +934,7 @@ def namespaced(namespace, new=True):
+    original_relative_names = cmds.namespace(query=True, relativeNames=True)
     if new:
         namespace = unique_namespace(namespace)
         cmds.namespace(add=namespace)
```
I think you're missing the lines that actually set the `relative_names` state before the context too.

This is what I shared with you on discord:
```python
@contextlib.contextmanager
def namespaced(namespace, new=True, relative_names=None):
    """Work inside namespace during context

    Args:
        new (bool): When enabled this will rename the namespace to a unique
            namespace if the input namespace already exists.
        relative_names (bool): Whether to force relative names to True or
            False. Defaults to None to not change the current state.

    Yields:
        str: The namespace that is used during the context

    """
    original = cmds.namespaceInfo(cur=True, absoluteName=True)
    original_relative_names = cmds.namespace(query=True, relativeNames=True)
    if new:
        namespace = unique_namespace(namespace)
        cmds.namespace(add=namespace)
    if relative_names is not None:
        cmds.namespace(relativeNames=relative_names)

    try:
        cmds.namespace(set=namespace)
        yield namespace
    finally:
        cmds.namespace(set=original)
        if relative_names is not None:
            cmds.namespace(relativeNames=original_relative_names)
```

Note how before the yield we're also applying the relative names.

BigRoy commented on openpype/hosts/maya/plugins/load/_load_animation.py:
```diff
@@ -1,4 +1,46 @@
+
+
+def _process_reference(file_url, name, namespace, options):
+    """_summary_
```
Summary seems WIP :)

moonyuet replied:
i have added the docstring back

BigRoy commented on openpype/hosts/maya/plugins/publish/validate_skeleton_top_group_hierarchy.py:
```diff
@@ -0,0 +1,40 @@
+    """
+    order = ValidateContentsOrder + 0.05
+    label = "Top Group Hierarchy"
+    families = ["rig"]
```
Should this target `rig.fbx` since it only applies if skeleton mesh was found?

BigRoy commented on openpype/hosts/maya/plugins/publish/validate_skeleton_rig_content.py:
```diff
@@ -0,0 +1,101 @@
+    hosts = ["maya"]
+    families = ["rig.fbx"]
+
+    accepted_output = ["mesh", "transform", "locator"]
```
This would be best as a `set` since we're doing lookups into it.
```suggestion
    accepted_output = {"mesh", "transform", "locator"}
```

BigRoy commented on openpype/hosts/maya/plugins/publish/validate_skeleton_rig_controller.py:
I know the duplication of these validators is due to me saying "this is likely best separated" but looking at how much code we're duplicating I wonder if we're better of putting this validator in the same file in the original.

Then in the original validator we do:

```python
class OriginalValidator(pyblish.api.InstancePlugin):
    @classmethod
    def get_invalid(cls, instance):
        nodes_to_check = cls.get_nodes(instance)
        # Here put all the validation log that's there - we just add the `nodes_to_check` filter
        ...

    @classmethod
    def get_nodes(cls, instance):
        return control_nodes


class SkeletonValidator(OriginalValidator):
    @classmethod
    def get_nodes(cls, instance):
        return skeleton_control_nodes
```

So we would still be creating the additional validator but won't have the need to duplicate A TON OF CODE like we have now.

A similar approach is [used here](https://github.com/ynput/OpenPype/blob/d3e43301ac39a31ad2ffbcda76b651d26d10ef27/openpype/hosts/maya/plugins/publish/collect_look.py#L623-L662).

What do you think?

moonyuet replied:
I think this is a good idea although it needs some time to refactor the thing.

BigRoy commented on openpype/hosts/maya/plugins/publish/validate_rig_contents.py:
```diff
@@ -144,7 +150,84 @@ def validate_controls(self, set_members):
         return invalid
+
+    @classmethod
+    def get_nodes(cls, instance):
```
The return values, what it should return, isn't obvious at first so it might be good to add a decent docstring for these methods.

BigRoy reviewed (commented):
Some first comments, will take a better look hopefully tomorrow.

moonyuet commented on openpype/hosts/maya/plugins/publish/validate_rig_contents.py:
```diff
@@ -144,7 +150,84 @@ def validate_controls(self, set_members):
+
+    accepted_output = {"mesh", "transform", "locator"}
+
+    @classmethod
```
`rig contents` validate two sets(`out_SET` and `controls_SET`) while `skeleton rig content` validates just one set (`skeletonMesh_SET`).  Maybe we can try to abstract some codes into functions which makes the duplicates less.

moonyuet commented on openpype/hosts/maya/plugins/publish/extract_fbx_animation.py:
```diff
@@ -0,0 +1,60 @@
+        # FBX does not include the namespace but preserves the node
+        # names as existing in the rig workfile
+        namespace, relative_out_set = out_set_name.split(":", 1)
+        cmds.namespace(relativeNames=True)
```
This raises `ValueError: No object matches name: {TheRigName}` once it is removed. The system can't find the name of object_sets without namespace for exporting.

BigRoy replied:
Actually, that means the context manager doesn't work the way we intended it to - since we added the `relative_names=True` flag which should take care of this. Thus the context manager is probably buggy.

[This](https://github.com/ynput/OpenPype/pull/5589/files#r1342476496) should fix that so that here you can do:
```suggestion
```

BigRoy commented on openpype/hosts/maya/api/lib.py:
```diff
@@ -934,15 +934,19 @@ def namespaced(namespace, new=True):
         namespace = unique_namespace(namespace)
         cmds.namespace(add=namespace)
-
+        if relative_names is not None:
+            cmds.namespace(relativeNames=relative_names)
```
```suggestion
    if relative_names is not None:
        cmds.namespace(relativeNames=relative_names)
```

BigRoy commented on openpype/hosts/maya/plugins/publish/validate_rig_out_set_node_ids.py:
```diff
@@ -85,3 +85,50 @@ def repair(cls, instance):
+
+    """
+
+    order = ValidateContentsOrder
+    families = ["rig.fbx"]
+    hosts = ['maya']
+    label = 'Skeleton Rig Out Set Node Ids'
+    actions = [
+        openpype.hosts.maya.api.action.SelectInvalidAction,
+        RepairAction
+    ]
+    allow_history_only = False
```
You should only redefine what you're actually changing. If the docstring should be the same, then remove that too. :) This goes for all the other plugins too that inherit like this.
```suggestion
    families = ["rig.fbx"]
    label = 'Skeleton Rig Out Set Node Ids'
```

moonyuet commented on openpype/hosts/maya/plugins/publish/collect_skeleton_mesh.py:
```diff
@@ -0,0 +1,43 @@
+                    self.log.debug(
+                        "Collected skeletonmesh Set: {}".format(
+                            skeleton_mesh_content
+                        ))
```
I actually find out that it still adds the `rig.fbx` family even though it doesn't have objects inside the `skeletonMesh_SET`.

BigRoy replied:
Ah that's correct. You wouldn't want that.
```suggestion
        skeleton_mesh_content = cmds.sets(skeleton_mesh_set, query=True)
        if skeleton_mesh_content:
            instance.data["families"] += ["rig.fbx"]
            instance.data["skeleton_mesh"] = skeleton_mesh_content
            self.log.debug(
                "Collected skeletonMesh_SET members: {}".format(
                    skeleton_mesh_content
                ))
```
Is that what you wanted?

BigRoy commented on openpype/hosts/maya/plugins/publish/validate_animated_reference.py:
```diff
@@ -0,0 +1,65 @@
+
+    @classmethod
+    def validate_controls(self, set_members):
+        """Check if the controller set passes the validations
+
+        Checks if all its set members are within the hierarchy of the root
+        Checks if the node types of the set members valid
```
That is validates controls is pretty clear from the function. it's best to describe _what_ it checks.

```suggestion
        """Check if the controller set contains only accepted node types.
```

BigRoy commented on openpype/hosts/maya/plugins/publish/validate_rig_contents.py:
```diff
@@ -61,46 +55,91 @@ def process(self, instance):
+
+    @classmethod
+    def invalid_hierarchy(cls, instance, content):
+        """Check if the rig sets passes the validation with
+        correct hierarchy
```
It's unclear from the docstring what that correct hierarchy would be.
```suggestion
        """Check if all rig set members are within the hierarchy of the rig root
```


Also, can we remove this line from the docstring at [line 145](https://github.com/ynput/OpenPype/pull/5589/files#diff-9695eda766fbdeb12ee27150fd675de9f4f8cc029d06fe98a03551b9c2f3d040R145). It states:

> Checks if all its set members are within the hierarchy of the root

But that function does not do that - it only validates the geometry.

BigRoy commented on openpype/hosts/maya/plugins/publish/validate_rig_contents.py:
```diff
@@ -122,12 +161,11 @@ def validate_geometry(self, set_members):
-    def validate_controls(self, set_members):
+    @classmethod
+    def validate_controls(cls, set_members):
         """Check if the controller set passes the validations
 
         Checks if all its set members are within the hierarchy of the root
```
```suggestion
        """Checks if the control set members are allowed node types.
```

BigRoy commented on openpype/hosts/maya/plugins/publish/collect_fbx_animation.py:
```diff
@@ -0,0 +1,36 @@
+        if not self.is_active(instance.data):
+            return
+        skeleton_sets = [
+            i for i in instance
+            if i.lower().endswith("skeletonanim_set")
```
Does the "lower" check really make sense if we're enforcing it everywhere to be exactly the name with the uppercase, etc.?

BigRoy commented on openpype/hosts/maya/plugins/publish/collect_skeleton_mesh.py:
```diff
@@ -0,0 +1,45 @@
+
+        instance.data["skeleton_mesh"] = []
+
+        if skeleton_mesh_set:
+            skeleton_mesh_content = cmds.sets(
+                skeleton_mesh_set, query=True) or []
+            if not skeleton_mesh_content:
+                self.log.debug(
+                    "No object nodes in skeletonMesh_SET. "
+                    "Skipping collecting of skeleton mesh..."
+                )
+                return
+            instance.data["families"] += ["rig.fbx"]
+            instance.data["skeleton_mesh"] = skeleton_mesh_content
+            self.log.debug(
+                "Collected skeletonMesh_SET members: {}".format(
+                    skeleton_mesh_content
+                ))
```
```suggestion
        skeleton_mesh_content = cmds.sets(
            skeleton_mesh_set, query=True) or []
        if not skeleton_mesh_content:
            self.log.debug(
                "No object nodes in skeletonMesh_SET. "
                "Skipping collecting of skeleton mesh..."
            )
            return
        instance.data["families"] += ["rig.fbx"]
        instance.data["skeleton_mesh"] = skeleton_mesh_content
        self.log.debug(
            "Collected skeletonMesh_SET members: {}".format(
                skeleton_mesh_content
            ))
```

BigRoy commented on openpype/hosts/maya/plugins/publish/extract_fbx_animation.py:
```diff
@@ -0,0 +1,63 @@
+        # Export from the rig's namespace so that the exported
+        # FBX does not include the namespace but preserves the node
+        # names as existing in the rig workfile
+        namespace, relative_out_set = out_set_name.split(":", 1)
+        with namespaced(
+            ":" + namespace,
+            new=False,
+            relative_names=True
+        ) as namespace:
+            fbx_exporter.export(relative_out_set, path)
```
A question - why do we need to set `relative_names=True`? Is that required for the FBX exporter to ignore the namespace?
Because if not, we could just set `relative_names=`False` and export with `out_set_name` instead of the relative name? :)

moonyuet replied:
We need to import the fbx animation into other dcc such as Unreal and some of them require the strict hierarchy names so that's why we need to ignore the namespace for this extractor by default. 
but we can do a setting for this and enabled by default. Once the users dont want to have the namespace, they can just disable it.

BigRoy replied:
I understand - what I meant was; that if you were to go into the namespace with relative names `False` does it still export without the namespace. Did a quick test run on my end with export FBX and it does seem to require the relative names to be enabled - so will mark this resolved.

BigRoy commented on openpype/hosts/maya/plugins/publish/extract_fbx_animation.py:
```diff
@@ -0,0 +1,63 @@
+
+        fbx_exporter.set_options_from_instance(instance)
+
+        out_set_name = next(out for out in out_set)
```
This is the same as `out_set[0]` - is that what it was intended to be? It's basically the first node member of the `skeletonAnim_SET`. Seems odd behavior? :)

moonyuet replied:
I am looking for the top/root node of the members.

BigRoy commented on openpype/hosts/maya/plugins/publish/extract_fbx_animation.py:
```diff
@@ -0,0 +1,63 @@
+        path = path.replace("\\", "/")
+
+        fbx_exporter = fbx.FBXExtractor(log=self.log)
+        out_set = instance.data.get("animated_skeleton", [])
```
Note that this is not the _set_ but `animated_skeleton` data holds the members of the set.

BigRoy commented on openpype/hosts/maya/plugins/publish/extract_fbx_animation.py:
```diff
@@ -0,0 +1,62 @@
+            new=False,
+            relative_names=True
+        ) as namespace:
+            fbx_exporter.export(relative_out_set, path)
```
We might need to implement these:
```python
def get_namespace(node):
    """Return namespace of given node"""
    return node.rsplit("|", 1)[-1].rsplit(":", 1)[0]


def strip_namespace(node, namespace):
    """Strip given namespace from node path.
    
    The namespace will only be stripped from names
    if it starts with that namespace. If the namespace
    occurs within another namespace it's not removed.
    
    Examples:
        >>> strip_namespace("namespace:node", namespace="namespace:")
        "node"
        >>> strip_namespace("hello:world:node", namespace="hello:world")
        "node"
        >>> strip_namespace("hello:world:node", namespace="hello")
        "world:node"
        >>> strip_namespace("hello:world:node", namespace="world")
        "hello:world:node"
        >>> strip_namespace("ns:group|ns:node", namespace="ns")
        "group|node"
        
    Returns:
        str: Node name without given starting namespace.
    
    """
    
    # Ensure namespace ends with `:`
    if not namespace.endswith(":"):
        namespace = "{}:".format(namespace)
    
    # The long path for a node can also have the namespace
    # in its parents so we need to remove it from each
    return "|".join(
        name[len(namespace):] if name.startswith(namespace) else name
        for name in node.split("|")
    )
```

And then ensure we get and strip correctly for the members we want:
```suggestion
        out_members = instance.data.get("animated_skeleton", [])
        # Export
        instance.data["constraints"] = True
        instance.data["skeletonDefinitions"] = True
        instance.data["referencedAssetsContent"] = True

        fbx_exporter.set_options_from_instance(instance)

        # Export from the rig's namespace so that the exported
        # FBX does not include the namespace but preserves the node
        # names as existing in the rig workfile
        namespace = get_namespace(out_members[0])
        relative_out_members = [
            strip_namespace(node, namespace) for node in out_members
        ]
        with namespaced(
            ":" + namespace,
            new=False,
            relative_names=True
        ) as namespace:
            fbx_exporter.export(relative_out_members, path)
```

BigRoy commented on openpype/hosts/maya/api/lib.py:
```diff
@@ -183,6 +183,47 @@ def maintained_selection():
             cmds.select(clear=True)
 
 
+def get_namespace(node):
+    """Return namespace of given node"""
+    return node.rsplit("|", 1)[-1].rsplit(":", 1)[0]
```
I think my original suggestion will fail to correctly report the namespace if the node had no namespace. Here's an improved version:
```suggestion
def get_namespace(node):
    """Return namespace of given node"""
    node_name = node.rsplit("|", 1)[-1]
    if ":" in node_name:
        return node_name.rsplit(":", 1)[0]
    else:
        return ""
```

Example test:
```python
def get_namespace(node):
    """Return namespace of given node"""
    node_name = node.rsplit("|", 1)[-1]
    if ":" in node_name:
        return node_name.rsplit(":", 1)[0]
    else:
        return ""
        
def get_namespace_old(node):
    """Return namespace of given node"""
    return node.rsplit("|", 1)[-1].rsplit(":", 1)[0]
        
print(get_namespace("A"))
# ""
print(get_namespace_old("A"))
# "A"
print(get_namespace("A:B:C"))
# "A:B"
print(get_namespace_old("A:B:C"))
# "A:B"
```

BigRoy reviewed (approved):
I've done quite a few code reviews of this one by now - sorry for the constant hassle, haha. For as far as I can see the code looks good.

@LiborBatek could you do another test run with all the latest code changes?

LiborBatek reviewed (approved):
All seems working normally as before code changes.

Did tested to produce the actual ```rig``` asset, then extract ```animation``` with and without ```Fbx``` representation and loading/manually import those back and all seems fine!

![image](https://github.com/ynput/OpenPype/assets/112623825/4dfa0568-2794-448e-b922-18ad3f484183)

BigRoy commented:
> **All seems working normally** as before code changes.
> 
> ...
> 
> ![image](https://user-images.githubusercontent.com/112623825/272293521-4dfa0568-2794-448e-b922-18ad3f484183.png)

Looking at that screenshot I wouldn't say _that_ looks normal. 🤣
